client/driver/docker_progress.go

```diff
@@ -0,0 +1,259 @@
+func (p *imageProgress) set(msg *jsonmessage.JSONMessage) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.lastMessage = msg
+	p.timestamp = time.Now()
+
+	lps := lpsFromString(msg.Status)
+	if lps == layerProgressStatusUnknown {
+		return
+	}
+
+	layer, ok := p.layers[msg.ID]
+	if !ok {
+		layer = &layerProgress{id: msg.ID}
+		p.layers[msg.ID] = layer
+	}
+	layer.status = lps
+	if msg.Progress != nil && lps == layerProgressStatusDownloading {
+		layer.currentBytes = msg.Progress.Current
+		layer.totalBytes = msg.Progress.Total
+	} else if lps == layerProgressStatusDownloaded {
+		layer.currentBytes = layer.totalBytes
+	}
+}
+
+// currentBytes iterates through all image layers and sums the total of
+// current bytes. The caller is responsible for acquiring a read lock on the
+// imageProgress struct
+func (p *imageProgress) currentBytes() int64 {
+	var b int64
+	for _, l := range p.layers {
+		b += l.currentBytes
+	}
+	return b
+}
+
+// totalBytes iterates through all image layers and sums the total of
+// total bytes. The caller is responsible for acquiring a read lock on the
+// imageProgress struct
+func (p *imageProgress) totalBytes() int64 {
+	var b int64
+	for _, l := range p.layers {
+		b += l.totalBytes
+	}
+	return b
+}
+
+// progressReporterFunc defines the method for handeling inactivity and report
+// events from the imageProgressManager. The image name, current status message,
+// timestamp of last received status update and timestamp of when the pull started
+// are passed in.
+type progressReporterFunc func(image string, msg string, timestamp time.Time, pullStart time.Time)
+
+// imageProgressManager tracks the progress of pulling a docker image from an
+// image repository.
+// It also implemented the io.Writer interface so as to be passed to the docker
+// client pull image method in order to receive status updates from the docker
+// engine api.
+type imageProgressManager struct {
+	imageProgress    *imageProgress
+	image            string
+	activityDeadline time.Duration
+	inactivityFunc   progressReporterFunc
+	reportInterval   time.Duration
+	reporter         progressReporterFunc
+	cancel           context.CancelFunc
+	stopCh           chan struct{}
+	buf              bytes.Buffer
+}
+
+func newImageProgressManager(
+	image string, cancel context.CancelFunc,
+	inactivityFunc, reporter progressReporterFunc) *imageProgressManager {
+
+	return &imageProgressManager{
+		image:            image,
+		activityDeadline: defaultPullActivityDeadline,
+		inactivityFunc:   inactivityFunc,
+		reportInterval:   defaultImageProgressReportInterval,
+		reporter:         reporter,
+		imageProgress: &imageProgress{
+			timestamp: time.Now(),
+			layers:    make(map[string]*layerProgress),
+		},
+		cancel: cancel,
+		stopCh: make(chan struct{}),
+	}
+}
+
+// start intiates the ticker to trigger the inactivity and reporter handlers
+func (pm *imageProgressManager) start() {
+	pm.imageProgress.pullStart = time.Now()
+	go func() {
```

Comment:
Any thoughts on having `newImageProgressManager` call `go pm.start()` and then the interface for callers just becomes new and stop?

Reply:
The only possible hiccup I see is that there are some fields that we don't set in the `newImageProgressManager`. They are currently set to default, ex: `reportInterval`, `activityDeadline`, but could be overridden after the new call and before `start`. If we ever want to set these I suppose we just add the args to the new call.